### PR TITLE
bugfix: add workdir support to core.#Copy and docker.#Copy

### LINKED ISSUE #2512: 🐞 workdir is ignored in most docker.#Copy
### What is the issue?

After setting a workdir on the image config the #Copy steps writes to '/'. My assumption is that the #Copy and other commands would follow Dockerfile type guidance and respect workdir automatically.

### Log output

4:17PM INF system | installing all packages...
4:17PM INF system | installed/updated package dagger.io@0.2.12
4:17PM INF system | installed/updated package universe.dagger.io@0.2.12

dagger do --plan workdir-copy.cue test --log-format=plain
4:22PM INF actions.test.copyData._build._dag."0"._op | computing
4:22PM INF client.filesystem.".".read | computing
4:22PM INF client.filesystem.".".read | completed    duration=100ms
4:22PM INF actions.test.copyData._testFile | computing
4:22PM INF actions.test.copyData._testFile | completed    duration=100ms
4:22PM INF actions.test.copyData._build._dag."0"._op | completed    duration=15.6s
4:22PM INF actions.test.copyData._build._dag."2"._copy | computing
4:22PM INF actions.test.copyData._build._dag."2"._copy | completed    duration=300ms
4:22PM INF actions.test.readFile._test | computing
4:22PM INF actions.test.readFileIncorrect._test | computing
4:22PM INF actions.test.readFileIncorrect._test | completed    duration=0s
4:22PM ERR actions.test.readFile._test | failed: ReadFile /opt/testfile.txt: open /tmp/buildkit-mount1037262984/opt/testfile.txt: no such file or directory
    duration=0s
4:22PM FTL system | failed to execute plan: task failed: actions.test.readFile._test: ReadFile /opt/testfile.txt: open /tmp/buildkit-mount1037262984/opt/testfile.txt: no such file or directory

### Steps to reproduce

**Steps:**
1. Create a docker.#Build.
2. Add a docker.#Config step setting workdir (`/opt`)
3. Add a docker.#Copy step, copying a file. (`testfile.txt`)
4. Try to read the file from the workdir + path (`/opt/testfile.txt`) of the file by accessing the image #FS.

**Expected outcome:**
File found at `/opt/testfile.txt`
**Actual outcome:**
File cannot be found on the path `/opt/testfile.txt`. It has been written to `/testfile.txt`

Example test:
```
package main

import (
	"dagger.io/dagger"
	"dagger.io/dagger/core"
	"universe.dagger.io/docker"
)

dagger.#Plan & {
	client: filesystem: ".": read: contents: dagger.#FS

	actions: test: {
		copyData: {
			_testFile: core.#WriteFile & {
				input: client.filesystem.".".read.contents
				path: "./testfile.txt"
				contents: "test"
			}

			_build: docker.#Build & {
				steps: [
					docker.#Pull & {
						source: "index.docker.io/debian"
					},
					docker.#Set & {
						config: workdir: "/opt"
					},
					docker.#Copy & {
						contents: _testFile.output
					},
				]
			}

			output: _build.output
		}

		readFileIncorrect: {
			_test: core.#ReadFile & {
				input: copyData.output.rootfs
				path: "/testfile.txt"
			}
		}

		readFile: {
			_test: core.#ReadFile & {
				input: copyData.output.rootfs
				path: "/opt/testfile.txt"
			}
		}

	}
}

```

### Dagger version

dagger 0.2.12 (21a15a84) darwin/amd64

### OS version

macOS 12.3.1

### PR DESCRIPTION
**Pull request by zach-source**
This PR adds support for workdir with the core.#Copy function. Closes #2512. 

## What's changed
* Added a new workdir field to core.#Copy
* Use the workdir field in `plan/task/copy.go` to set the inputState.Dir to the working directory.
* docker.#Copy sets the workdir field based on the image.config.workdir
*Test cases to verify copying with workdir, without workdir, and with workdir and absolute value.

I played with just cuelang modifications in the docker/build.cue file but had to use path.Join and wasn't a good way to get the os type so abandoned that for this approach.

### CODE REVIEW

File: pkg/dagger.io/dagger/core/fs.cue [hunk 1]
```diff
@@ -72,10 +72,12 @@ import "dagger.io/dagger"
 	input: dagger.#FS
 	// Contents to copy
 	contents: dagger.#FS
+	// Working dir (optional)
+	workdir: string | *"/"
```
Comment (sipsma) [inline pkg/dagger.io/dagger/core/fs.cue:76]: Saw your comment about which OS to use for operations in the `path` package. I think we can safely assume `path.Unix`. Even though we support Dagger clients running on many platforms, AFAIK all dagger actions can only be expected to run with unix filepaths.

I think that would be a better approach overall too by keeping our `core` interface as minimal as possible; seems like it should allow the change exist only in `docker/build.cue`.

**Review by sipsma — COMMENTED** [review]
Left a comment, but looks good overall and appreciate the PR! The change makes sense to me but I'll defer to others with more context on the `docker` package to confirm this is the behavior we'd want.

**Review by shykes — DISMISSED** [review]
Thanks for the contribution Zachary.

Adding workdir support to `docker.#Copy` makes sense, as made clear by #2512. However, this support should not leak into `core.#Copy`.
* The "sticky defaults" behavior of Dockerfiles is implemented *in pure CUE* in the `docker` package. This allows us to keep core operations like `#Copy` or `#Exec` simpler, because there is no parallel configuration trapped in the llb state black box.
* This PR as proposed would break that design

Basically I recommend solving #2512 purely in `docker.#Copy` and not touching core for this.

Thank you again for the contribution and sorry for the disappointment / extra work.

File: pkg/dagger.io/dagger/core/fs.cue (hunk 1 above)
Comment (zach-source) [reply pkg/dagger.io/dagger/core/fs.cue:76]: Awesome, that's exactly my follow up question. I'll rework it and resubmit with that.

**Comment by shykes** [comment]
> Not for this bug but I feel like bash.#Run should also support workdir. Do you agree with that? I can open a separate bug to track and fix there as well.

`bash.#Run` wraps `docker.#Run` and therefore already supports workdir 😅

**Comment by zach-source** [comment]
@shykes Hrm, that's not the behavior I'm seeing. I'll try to replicate it.

**Comment by zach-source** [comment]
@shykes Must have been something weird with my tests, independent tests show it works fine. I did notice a bug with `export: files` not using workdir so you have to use absolute paths but the script and files are written to the correct dir.

**Review by sipsma — APPROVED** [review]
Updates LGTM, thanks again for the fix!

**Review by TomChv — APPROVED** [review]
LGTM

**Comment by sipsma** [comment]
The request from @shykes to avoid changes to `core` tasks has been addressed, so I'm going to dismiss that change request and merge this as is given there's two approvals.
